sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc

```diff
@@ -0,0 +1,510 @@
+used by permission by Khronos.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to
+a feature for review and community feedback. When the feature matures, this
+specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are
+subject to change they are not intended to be used by shipping software
+products.
+
+== Version
+
+Built On: {docdate} +
+Revision: 2
+
+== Contact
+
+Abhishek Tiwari, Intel (abhishek2 'dot' tiwari 'at' intel 'dot' com)
+
+== Contributors
+
+Abhishek Tiwari, Intel +
+Gregory Lueck, Intel +
+Jason Sewall, Intel +
+Jessica Davies, Intel +
+Joe Garvey, Intel +
+John Pennycook, Intel +
+Michael Kinsner, Intel +
+Roland Schulz, Intel
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, revision 4.
+
+It also depends on the `sycl_ext_oneapi_properties` extension.
+
+== Overview
+
+[NOTE]
+====
+In this document, we use the shortened form `annotated_ptr` to refer to the
+proposed `sycl::ext::oneapi::annotated_ptr` class.
+====
+
+The purpose of this document is to clearly describe and specify `annotated_ptr`
+and related concepts, types, and mechanisms, and to give examples and context
+for their usage.
+
+=== Motivation
+
+When compiling a program with pointer operations, it is often desirable to
+attach certain information such as optimization hints or functional directives
+onto the pointer itself.
+
+Any acceptable mechanism to attach such information should meet the following requirements:
+  1. It should not be burdensome for users to use the mechanism.
+  2. It should be reasonably scalable and not add significant burden on compiler developers to maintain the mechanism.
+  3. It should preserve the attached information through most uses of the pointer.
+
+For example, consider the following case: We want to specify some compile-time constant information `I` for a pointer. One possible method to do so is for the compiler to provide a special function for reading to and writing from the pointer and have this function convey `I` to the compiler at the call site where the read or write occurs.
+
+That is, users would convert code that looks like:
+```c++
+{
+  SomeType* a;
+  ...
+  *a = some_val;
+  ...
+  read_val = *a
+}
+```
+
+To code that looks like:
+```c++
+{
+  SomeType* a;
+  ...
+  __special_store_with_annotation(a, some_val, I /*some compile time constant information*/);
+  ...
+  __special_load_with_annotation(a, &read_val, I /*some compile time constant information*/);
+}
+```
+
+This mechanism does not meet requirements (1) and (2) listed above because:
+
+  * Users have to replace all their pointer read/write code with special function calls.
+  * If the set or type of information represented by `I` evolves over time, it requires creating and/or updating these special functions and the underlying compiler implementation.
```

Comment:
I'm not convinced this is a lot of work to maintain.  You'd have to do some work in the compiler optimizer or backend to parse the new information but your frontend probably doesn't have to change if __special_load_with_annotation and the format of I is defined carefully for extensibility.  I think your first argument, above, is more convincing and alone is enough to disqualify this approach.  I'd drop this second, weaker argument.

Reply:
Updated.